examples/imatrix/imatrix.cpp

```diff
@@ -252,102 +268,136 @@ void IMatrixCollector::save_imatrix(int ncall) const {
+            struct ggml_tensor * counts = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, 1, nmat);
+            ggml_set_name(sums, (name + ".sums").c_str());
+            ggml_set_name(counts, (name + ".counts").c_str());
+
+            for (int32_t j = 0; j < nval; ++j) {
+                ((float *) sums->data)[j] = (float) stat.values[j];
             }
-            out.write((const char*)tmp.data(), nval*sizeof(float));
+            for (int32_t j = 0; j < nmat; ++j) {
+                ((float *) counts->data)[j] = (float) stat.counts[j];
+            }
+
+            gguf_add_tensor(ctx_gguf, sums);
+            gguf_add_tensor(ctx_gguf, counts);
         }
     }
 
-    // Write the number of call the matrix was computed with
-    out.write((const char *) &m_last_call, sizeof(m_last_call));
-
-    // Write the input filename at the end of the file to later on specify it in quantize
-    {
-        int len = m_params.prompt_file.size();
-        out.write((const char *) &len, sizeof(len));
-        out.write(m_params.prompt_file.c_str(), len);
-    }
+    gguf_write_to_file(ctx_gguf, fname.c_str(), false);
 
     if (m_params.verbosity > 0) {
-        fprintf(stderr, "\n%s: stored collected data after %d chunks in %s\n", __func__, m_last_call, fname.c_str());
+        fprintf(stderr, "\n%s: stored collected data after %d chunks in %s\n", __func__, m_last_chunk, fname.c_str());
     }
+
+    gguf_free(ctx_gguf);
+    ggml_free(ctx);
 }
 
-bool IMatrixCollector::load_imatrix(const char * fname) {
-    std::ifstream in(fname, std::ios::binary);
-    if (!in) {
-        printf("%s: failed to open %s\n",__func__, fname);
+bool IMatrixCollector::load_imatrix(const char * file_name) {
+    struct ggml_context * ctx = nullptr;
+    struct gguf_init_params meta_gguf_params = {
+        /* .no_alloc = */ false, // the data is needed
+        /* .ctx      = */ &ctx,
+    };
+    struct gguf_context * ctx_gguf = gguf_init_from_file(file_name, meta_gguf_params);
+    if (!ctx_gguf) {
         return false;
     }
-    int n_entries;
-    in.read((char*)&n_entries, sizeof(n_entries));
-    if (in.fail() || n_entries < 1) {
-        printf("%s: no data in file %s\n", __func__, fname);
+    const int32_t n_entries = gguf_get_n_tensors(ctx_gguf);
+    if (n_entries < 2) {
+        fprintf(stderr, "%s: no data in file %s\n", __func__, file_name);
+        gguf_free(ctx_gguf);
+        ggml_free(ctx);
         return false;
     }
-    for (int i = 0; i < n_entries; ++i) {
-        int len; in.read((char *)&len, sizeof(len));
-        std::vector<char> name_as_vec(len+1);
-        in.read((char *)name_as_vec.data(), len);
-        if (in.fail()) {
-            printf("%s: failed reading name for entry %d from %s\n",__func__,i+1, fname);
+
+    const std::string sums_suffix{".sums"};
+    const std::string counts_suffix{".counts"};
+
+    // TODO: allow loading from mis-ordered imatrix files
```

Comment:
Good idea. I've implemented this in <https://github.com/ggerganov/llama.cpp/pull/9400/commits/8c13e16bb0f6b654ff4774e54fbc3b125ae495a6> by using a single `std::map<std::string, std::pair<struct ggml_tensor *, struct ggml_tensor *>>` and using the common prefix as the `name`, because the sums and counts tensors are iterated simultaneously in `quantize.cpp`, and this allows using a similar implementation in both `load_imatrix`.